typescript-test-samples/apigw-lambda-external/README.md

```diff
@@ -0,0 +1,119 @@
+
+### System Under Test (SUT)
+
+The SUT in this pattern is a synchronous API composed of Amazon API Gateway, AWS Lambda and a simple external API.
+
+![System Under Test (SUT)](img/system-under-test.png)
+
+### Goal
+
+The goal of this pattern is to test the SUT in an environment as similar to production as possible by running tests against resources deployed to the cloud but not communicating with an external API.
+
+### Description
+
+The SUT will be deployed to the AWS cloud. The test code will create an HTTP client that will send requests to the deployed API Gateway endpoint. The endpoint will invoke the backing services, test resource configuration, IAM permissions, authorizers, and internal business logic of the SUT. The SUT will not have any dependency on the external API, resulting in the service not being negatively impacted by the constraints or idiosyncrasies of external systems, ensuring that this system remains robust whilst under test.
+
+![System Under Test Description (SUT)](img/system-under-test-description.png)
+
+## [Top](#contents)
+
+## About this Example
+
+This sample project allows a user to call an API endpoint with a city location to retrieve the weather and multi day forecast using an external api [https://goweather.herokuapp.com/weather](https://goweather.herokuapp.com/weather). In this example the AWS API Gateway and AWS Lambda are acting as a proxy and providing no specific business logic over the response from the weather API.
+
+This project consists of an [API Gateway](https://aws.amazon.com/api-gateway/), a single [AWS Lambda](https://aws.amazon.com/lambda) function, and an [API Gateway](https://aws.amazon.com/api-gateway/) mock of the external dependency.
+
+### Key Files in the Project
+
+- [app.ts](src/app.ts) - Lambda handler code to test
+- [template.yaml](template.yaml) - SAM script for deployment
+- [test-handler.test.ts](src/tests/unit/test-handler.test.ts) - Unit test using mocks
+- [integration-handler.test.ts](src/tests/integration/integration-handler.test.ts) - Integration tests on a live stack
+
+## [Top](#contents)
+
+## Run the Unit Tests
+
+[test-handler.test.ts](src/tests/unit/test-handler.test.ts)
+
+In the [unit test](src/tests/unit/test-handler.test.ts#L44), all calls to the external dependency are mocked.
+To run the unit tests
+
+```shell
+apigw-lambda-external$ cd src
+src $ npm install
+src $ npm run test:unit
+```
+
+[Top](#contents)
+
+---
+
+## Run the Integration Tests
+
+[integration-handler.test.ts](src/tests/integration/integration-handler.test.ts)
+
+As this sample contains reference to an external API, to perform the integration tests without the dependency a volatile resource [API Gateway](https://aws.amazon.com/api-gateway/) is required to behave as the external API. To achieve this an additional parameter `IsVolatile` is required during the `sam deploy` step. This will instruct the cloudformation to deploy the mock api and configure the Lambda to utilise this API Gateway.
```

Comment:
Mention they'll be prompted to set the parameter during `sam deploy --guided`.

Link to SAM docs on parameters.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html

Reply:
I've expanded on the two different ways to define the require parameter.

Note: 
I tried to include `AllowedValues` in the Cloudformation template, though didn't appear to receive any validation failures upon submitting a different value 🤔. 

I'm also uncertain that when deploying `true` to deploy the mock stack, when submitting again with `false` to the same stack, the mock did not appear to be removed, though the Lambda environment variable was updated 🤔.

Reply:
Thanks for the update!

I'll check what my stack does on the true/false flip too. Based on my reading of the [CloudFormation Conditions docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/conditions-section-structure.html) (last paragraph of the intro), it should be dropped.

Reply:
Confirming that with the Condition fixes, the Mock API is deleted as expected.